Extends set_label_position to position relative to axes or spine

ccaprani opened the pull request:
## PR Summary

This PR addresses https://github.com/matplotlib/matplotlib/issues/21333 and extends the axis label positions with `'axesleft'`, `'axesright'`, `'axesbottom'`, and `'axestop'`.

![image](https://user-images.githubusercontent.com/31228546/136810661-eea64d61-6ffa-469a-8e8c-43a8f340cd27.png)

## PR Checklist

<!-- Please mark any checkboxes that do not apply to this PR as [N/A]. -->

- [N/A] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [x] New features are documented, with examples if plot related.
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [x] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [N/A] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [N/A] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Do not create the PR out of master, but out of a separate branch.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

ccaprani commented:
Hi @jklymak, what are next steps for this PR? Documentation? And what release could we target? Thanks!

jklymak commented:
Yep tests and documentation so we can be sure it is working.  Thanks!

ccaprani commented:
Hi @jklymak ! Just a prompt on closing this one out please. Thanks.

jklymak commented:
@ccaprani The ball is in your court to fix the tests and the rebase, I think...

melissawm commented:
Hi @ccaprani - can we help you move this forward? Thanks!

ccaprani commented:
Hi @melissawm - thanks for the prompt. I've rebased now and hope it's ok. I think my auto-black-formatting on save has caused a few unintended formatting changes. Not quite sure how to undo these, but the intended changes are now clear - `axestop` and `axesbottom` keywords for positioning labels.

Would you kindly have a look and let me know if there is anything else needed?

jklymak commented:
You have to undo the black formatting!  Suggest you save the changes you made in another file, and reset your branch to upstream (`git reset --hard upstream/main`) and then re-add just the relevant changes and commit.

ccaprani commented:
Thanks for the suggestion @jklymak . I've done that now and also fixed the linting @melissawm . I think it's close now, but please let me know of anything else.

jklymak commented on examples/text_labels_and_annotations/align_axes_labels.py:
```diff
@@ -0,0 +1,101 @@
+    ax.set_xlabel(r'x-axis label')
+    ax.set_ylabel(r'y-axis label')
+
+fig = plt.figure(constrained_layout=True, figsize=(8, 8))
```
```suggestion
fig = plt.figure(figsize=(8, 8), layout='constrained')
```

jklymak commented on examples/text_labels_and_annotations/align_axes_labels.py:
```diff
@@ -0,0 +1,101 @@
+ax.xaxis.set_label_position('axestop')
+ax.yaxis.set_label_position('axesright')
+
+fig.set_tight_layout(True)
```
```suggestion
```
This undoes layout='constrained' and is incompatible with subfigures.

ccaprani replied:
Thank you!

jklymak commented on examples/text_labels_and_annotations/align_axes_labels.py:
```diff
@@ -0,0 +1,101 @@
+"""
+=================
+Align axes labels
+=================
```
```suggestion
==========================
Axes labels outside floating spines
==========================
```
So far as I can tell, this is not aligning the axes labels, but rather placing them definitively outside the spines

jklymak commented on examples/text_labels_and_annotations/align_axes_labels.py:
```diff
@@ -0,0 +1,101 @@
+    ax.set_ylabel(r'y-axis label')
+
+fig = plt.figure(constrained_layout=True, figsize=(8, 8))
+fig.suptitle('Aligning Axes Labels')
```
```suggestion
fig.suptitle('Axes Labels outside spines')
```

jklymak commented on lib/matplotlib/axis.py:
```diff
@@ -2584,7 +2605,21 @@ def _update_label_position(self, renderer):
             )
+        elif self.label_position == 'axesleft':
+            spinebbox = self.axes.bbox
+            bbox = mtransforms.Bbox.union(bboxes + [spinebbox])
```
Do you have a test of this if the spines are not floating?

ccaprani replied:
I've added a test now for normal axes; and thank you - because in doing so I did discover a little bug. Now the new keywords behave as a superset of the usual `top`, `left`, etc.

jklymak reviewed (commented):
I think this is close, but the example is not quite correct and testing should maybe include a how this looks on a normal axes...

ccaprani commented:
Hi @jklymak , one test is failing because I changed the baseline image. I presume this is not a problem?

jklymak commented:
@ccaprani you will need to replace the test image with the new one.

ccaprani commented:
Hi @jklymak I did change the baseline test image with the new one, i.e. [this one](https://github.com/matplotlib/matplotlib/pull/21338/files#diff-d722ce45a766909308782bcfefc6439a95bcc8a3dfa45e0d223a9085cffe04b9).

jklymak commented:
@ccaprani all the tests need to pass, so you still have some work to do!  Thanks,

ccaprani commented:
Hi @jklymak I completely agree that tests should pass, and I'm sorry I haven't adequately explained the problem.

The test it is failing is `PR cleanliness / pr_clean (pull_request)`, which reports:

```
The following images were both added and modified in this PR:
lib/matplotlib/tests/baseline_images/test_axes/axes_labels_position.png
Error: Process completed with exit code 1.
```
Now this image was (of course) changed/added in this PR as we discussed above because of the new test we have added.

So unless there is something I am not seeing, there is a catch-22: don't use a new test and pass the PR cleanliness test (poor); or introduce the new figure to test against and fail PR cleanliness (poor?). I'd appreciate if you can advise the way forward.

jklymak commented:
The PR cleanliness issue is because you have 4 commits that (presumably) swap between the files being there and not being there.  Once you squash the commits, the file thrashing will no longer be there and the PR will be "clean".  The purpose of this check is to make sure people don't commit big files into the repository that are then immediately removed, hence making the repo size grow for no reason.

melissawm commented:
@ccaprani if you need guidance on squashing, [GitHub has some nice docs on it](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/about-pull-request-merges#squash-and-merge-your-commits) 😄

ccaprani commented:
Thanks @melissawm : it seems the squash and merge feature of github requires write access to the repository, which of course I don't have. At least on repos I manage, I can only squash and merge a PR as the owner.

On the other hand, following @jklymak 's suggestion, I can [squash the commits into one](https://stackoverflow.com/questions/25356810/git-how-to-squash-all-commits-on-branch):
```
git reset --soft main
git add ...
git commit -m "All commits now as one"
```
but this then lists an enormous amount of files as part of the commit, which I think we also don't want.

Finally, presumably, the history on the branch is important too - or are we happy to overwrite that?

rcomer commented:
Hi @ccaprani, it looks like your new image test is failing.  Did you change the code since you generated the image?  If so, you will need to re-generate the image - [triage_tests.py](https://github.com/matplotlib/matplotlib/blob/main/tools/triage_tests.py) might help you with that.  If you commit with the `--amend` option, that will save you from needing to squash again.

When you're ready, please hit the "Ready for Review" button.  PRs marked as draft tend to be ignored.

rcomer commented:
Hi @ccaprani are you still interested in working on this?  Can we help you to move it forward?

ccaprani commented:
Thanks. I'd love if you could. I can't really see why it's failing when it's changing the test itself.

rcomer commented:
I ran the tests locally and got this for your test:

![axes_labels_position](https://github.com/matplotlib/matplotlib/assets/10599679/2a600eff-4696-4108-9e5c-6f5ef9ce102d)

So it looks like you added `remove_text=True` into the test decorator after you originally created the image.  The way to fix the test is therefore simply replace the reference image with the one that is produced now.

rcomer commented:
@ccaprani did you want to try replacing the test image, or would you prefer someone else to take over this work?

ccaprani commented:
Thanks. I'd like to finish it. Next few days. I've been travelling. Cheers!

jklymak commented on lib/matplotlib/tests/test_axes.py:
```diff
@@ -7902,6 +7902,89 @@ def test_ylabel_ha_with_position(ha):
     assert ax.yaxis.get_label().get_ha() == ha
 
 
+@image_comparison(baseline_images=['axes_labels_position'], remove_text=True,
```
I don't think you want to remove the text from this test

Also why are you doing this w a bunch of subfigures?  Seems un necessary and it's usually better not to cross test features in this manner.

ccaprani replied:
Thanks @jklymak - should the plots then all be separate? Subfigures seemed the easiest way to get it all in one image for testing.

jklymak replied:
Just use subplots 

I hadn't notice that you'd removed all ticks. In that case I guess you can remove all text too. Otoh it makes the tests harder to parse.

ccaprani commented:
Hi all,

Have rebased onto latest main, updated the `remove_text`, and changed the test to use subplots instead of subfigures.

The test is failing but I cannot explain it. The test function produces the expected image, but when called through pytest (i.e. wrapped with decorator), it seems the `constrained_layout` is not executed somehow, and so the test fails. I cannot understand why the decorator affects the subplot layouts.

![image](https://github.com/matplotlib/matplotlib/assets/31228546/7b26b7c4-9f43-4b46-91e1-07e159eebc3f)

Perhaps obvious to say, but once I use the pytest generated image as the expected image then it does pass; it's just that the test function when not decorated does not produce the 'expected' image then!

ccaprani commented:
@jklymak @rcomer gentle nudge on this to close it out - thanks!

jklymak commented:
@ccaprani the tests aren't passing, so you'll need to sort that out.  You need to generate new images that pass and commit them.

ksunden commented:
There are some differences in default rcParam values for test images compared to just using the library, so it is expected that you _do_ need to use the image created from running pytest.

One change that gets more similar images (though I believe at least the dpi is still different) is to pass `style="mpl20"` to the image comparison decorator.

By default it uses the "classic" (pre-mpl 2.0) style, which is done so that tests which predate that transition did not have to be recreated. New tests prefer using the new style. (We are working towards solutions to some of these problems) This is also why it uses bright blue instead of our more subdued blue for the lines.


ref: https://matplotlib.org/devdocs/devel/testing.html#writing-an-image-comparison-test

ccaprani commented:
Hi @jklymak, @ksunden ,

Would you be able to help get this over the line please? I think it's a useful addition.
